### PR TITLE
[IGNITE-20504] Sql. Improve test coverage of MappingService

### PR DESCRIPTION
**Pull request by lowka**
Adds tests for MappingService.

- `FragmentMappingTest` - includes test cases (uses `MappingTestRunner`).
- `MappingTestRunner` - loads a test file and executes test cases.
- `FragmentPrinter` - converts `MappedFragments` to text representation (used by `MappingTestRunner`).

Test files are located in `src/test/resources/mapping`. The tests use distributions that are used in production code (affinity, single, identity). Although tests do not use systems view, tests cluster-wide system view and node system views are replaced with tables that use aforementioned distributions.

To generate results for a test case use `MAPPING_TESTS_OVERWRITE_RESULTS`

_Symbol `\` should not be present in files it is only added because otherwise it breaks GitHub formatting._

> file: dummy.test
  N1
  SELECT 1
  \ ---
  f
  \ ---

>  @ Test
  @ WithSystemProperty(key = "MAPPING_TESTS_OVERWRITE_RESULTS", value = "true")
    public void testDummy() {
        addNodes("N0", "N1", "N2");
        addTable("T", "N1");
        testRunner.runTest(this::initSchema, "dummy.test");
    }

Updated test case:

> file: dummy.test
N1
SELECT 1
\---
Fragment#0 root
  nodes: [N1]
  tree:
    IgniteValues
\---

To strip results from all test cases in a file use `MAPPING_TESTS_STRIP_RESULTS` property (see test cases in `MappingTestRunnerSelfTest`).

https://issues.apache.org/jira/browse/IGNITE-20504

----

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentMappingTest.java [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+/**
+ * Tests for fragment mapping.
+ *
+ * <p>Test setup:
+ * <pre>
+ *     // add nodes named n1, n2, and n3 to the cluster.
+ *     addNodes("N1", "N2", "N3");
+ *
+ *     // The first statement adds a table with data located at node n1.
+ *     // The first one adds a table with data located at node n2, n3.
```
Comment (zstan) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentMappingTest.java:60]:
```suggestion
 *     // The second one adds a table with data located at node n2, n3.
```

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentMappingTest.java [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+ *     // The catalog is going to contain to tables "t1_n1" and "t2_n2n3".
```
Comment (zstan) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentMappingTest.java:61]:
```suggestion
 *     // The catalog is going to contain tables "t1_n1" and "t2_n2n3".
```
you need to change tables names below or change it here, as i understand

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentMappingTest.java [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+        addNodes("N1", "N2");
+
+        testRunner.runTest(this::initSchema, "values.test");
+    }
+
+    @Test
+    public void testTable() {
+        addNodes("N1", "N2");
+
+        addTable("T1", "N1");
```
Comment (zstan) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentMappingTest.java:107]: i try to play with this framework and change :
addNodes("N2", "N3");
addTable("T1", "N2");
and in the table_affinity.test i change :
N1
SELECT * FROM t1_n2
obtain test infinetely hanging

Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentMappingTest.java:109]: Fixed.

**Comment by lowka** [comment]
Fixed disabled rules hint propagation caused by a switch from `TestCluster` to `PlannerTests`.

**Comment by lowka** [comment]
Updated output format

> Fragment#N
   ..targetNodes: [nodes]
   ..executioNodes: [nodes]
  ..remoteFragments: [...]
  ..tables: [tableNames]
  ..systemViews: [systemViewNames]

Added column names in affinity distribution format.
Removed stable id generation (since this is no longer needed).

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentPrinter.java [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+                                        "No id for target source id=" + sourceId + ". Existing: " + collector.objectIndices);
+                            }
+                            return v;
+                        }
+                    })
+                    .sorted(Comparator.comparing(a -> a.format()))
+                    .collect(Collectors.toList());
+
+            output.appendPadding();
+            output.writeKeyValues("sources", sortedTargetIds);
```
[the remark this answers is not in the document]
Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentPrinter.java]: Fixed.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentPrinter.java [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+            if (wasBlank && newLinePadding > 0) {
+                builder.append(paddingString.repeat(newLinePadding));
+            }
+
+            if (wasBlank) {
+                blankLine = false;
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/FragmentPrinter.java]: Fixed.

**Review by korlov42 — APPROVED** [review]

**Review by zstan — APPROVED** [review]
LGTM, 1 minor comment, but it up to you.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/MappingTestRunner.java [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+        runTestCases(testFile, setup.schema, setup.executionTargetProvider, setup.topologySnapshot, parseValidate, testCases);
+    }
+
+    @TestOnly
+    void initFlags() {
+        overwriteResults = IgniteSystemProperties.getBoolean(OVERWRITE_RESULTS, false);
+        stripResults = IgniteSystemProperties.getBoolean(STRIP_RESULTS, false);
+
+        if (overwriteResults && stripResults) {
+            throw new IllegalStateException("Both overwriteResults and stripResults have been specified");
```
[the remark this answers is not in the document]
Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/mapping/MappingTestRunner.java]: fixed

**Pull request merged**